### ISSUE TITLE
Replace hard coded CSS classnames in base login theme with properties

### ISSUE THREAD

**Issue by dasniko** [keycloak/keycloak#30192]
### Description

I recognized that there are several CSS classnames hard coded in `base` login theme. Some are general like `checkbox`, some are PatternFly classes, like `pf-...`.

IMHO the `base` theme should be completely unstyled and should not contain any hard coded CSS classnames at all. Instead, replace them with properties (like most of the other classnames are being done) and set these properties with the default values in `keycloak` theme.

WDYT @ssilvert ?
Is this already on your agenda while creating the `keycloak.v2` login theme?

If not, I'd volunteer by providing a PR (might take some time, yes, but I'd do it).

### Motivation

Custom login themes inheriting from `base` would not have to deal with hardcoded classnames and could define custom classes easier without needing to adjust Freemarker files.

**Comment by ssilvert** [comment]
@dasniko In general, I agree that the class names should not be hard-coded in the *.ftl files.

For this enhancement, I'd rather that we concentrate on the `keycloak.v2`.  The old one will soon be deprecated.

For now, `keycloak.v2` login theme lives here: https://github.com/keycloak/keycloak/tree/main/themes/src/main/resources/theme/keycloak.v2/login

The `theme.properties` shows `keycloak` as its parent, but I'm not sure what it is using from there or from `base`.  I'll ask the rest of the UI team about it and see what they think.  Maybe we need a new `base` for `keycloak.v2`?

**Comment by jonkoops** [comment]
IMHO we should get rid of the base themes and move to them to a quick-start as a standalone theme that is not based on any sort of UI framework but just very barebones styling. Also re-usable page elements such as buttons, form elements, etc. should have their own re-usable templates so they can more easily be overwritten without needing to customize multiple templates.

**Comment by dasniko** [comment]
Thanks for your answers, @ssilvert and @jonkoops 
I totally understand that you want to concentrate on `keycloak.v2`. But also `v2` will have to inherit something, or will it be built-up completely on its own? Currently, it is based on `keycloak`, which is based on `base` - still the existing `base` theme.

I like the idea of having a completely loosely coupled `base` theme for building custom themes, it might make it easier to customize the look'n'feel when _not_ having to deal with PF-specific things. On the other hand side, if it's done as a starter, like @jonkoops mentioned, I see the danger that it will miss new features, as they only might be introduced to the (obvious) `keycloak.v2` theme in first place.

IMHO there should (must!?) be a completely unstyled base theme, providing all the functionality to the users/customers in a proper way.

Feel free to convert this to a discussion, rather than having this already as an issue. WDYT?

**Comment by jonkoops** [comment]
> But also `v2` will have to inherit something, or will it be built-up completely on its own?

Yeah, so this is currently extending `keycloak` as we wanted to port stuff over incrementally and fall back to the V1 theme where needed whilst we worked on implementation. Now that this is done, we need to start de-coupling the V2 theme from the `keycloak` theme, so it would start inhering from the `base` theme instead.

> IMHO there should (must!?) be a completely unstyled base theme, providing all the functionality to the users/customers in a proper way.

Yeah, that is the question. If we were to do this, we'd have to dogfood this for sure and also base our themes on this. Either way we'll be forced to make stack choices for the base theme due to the increasing amount of dynamic JavaScript on pages, which is not really manageable in just vanilla JS anymore. Hence we've been leaning towards Apline.js for the new login theme.

That in itself is not a reason to drop using the `base` theme, and we could keep supporting it. But the way the V2 theme is build up now it's quite different in structure due to all the changes in PatternFly between version 3 and 5 (and soon 6). So the base theme would need to start using a more component based architecture to remain relevant.

I don't see HTML with some properties sprinkled over it as a long-term scalable architecture, mostly because it's been proven internally with the PatternFly migration that the primitives here are too low-level to easily swap out a UI framework, which is likely what the community would want to do for their customization.

Another notable thing is that a lot of innovation has been happening with integrating Keycloak themes with existing front-end ecosystems, for example [Keycloakify](https://www.keycloakify.dev/) for React users. We might want to focus on what those communities need to build better abstractions so Keycloak becomes easier to integrate there.

My vote would be to get rid of the `base` themes entirely and to build out a starter template for Keycloak theme development that is pretty much basic plain HTML that users can fork to build their own, which is not to 'batteries included'. Additionally I would want to invest in guides to community maintained solutions (such as Keycloakify) for specific ecosystems such as React.

**Comment by edewit** [comment]
Actually I don't see the problem. When things already have class names you can still redefine those classes regardless of what the name is. For example when a text has `pf-v5-c-form__label-text` I can add define that as red text with black background. Even if I add patternfly css I can override the variables to completely make it my own. No need to change any freemarker template

**Comment by garronej** [comment]
> Another notable thing is that a lot of innovation has been happening with integrating Keycloak themes with existing front-end ecosystems, for example [Keycloakify](https://www.keycloakify.dev/) for React users. We might want to focus on what those communities need to build better abstractions so Keycloak becomes easier to integrate there.  

Thank you for recognizing us as part of the Keycloak community!

Regarding the issue at hand, I agree with @dasniko that it would be beneficial to avoid having hardcoded PatternFly classes in the base theme. The current approach base theme/keycloak theme is very effective.  

In Keycloakify, the customization strategy is as follows:
- For light configurations, users can target the 'kc' CSS class to apply their custom styles using plain CSS, Tailwind, or a CSS-in-JS solution. They benefit from type safety on the class names, making the results easy to maintain and providing great outcomes with minimal investment.
- When CSS-level customization is insufficient, users can "eject" some templates from the base theme that gets copied as react components into their project, enabling them to implement component-level customizations with component libraries such as ShadeCN UI or MUI.

It's often enough to just eject the template that is common across all pages.

If customization reaches a point where the theme no longer requires PatternFly, users can simply toggle a switch to disable it. This makes the presence of residual 'pf-' classes in the code undesirable.

I will link a tutorial video here that I'm currently re-recording in the next few days to illustrate these points.

On a side note, @kathari00 has successfully implemented Keycloakify in an Angular project and has even initiated a Keycloakify-starter-angular project: https://github.com/kathari00/keycloakify-starter-angular

Things are progressing quickly!

**Comment by ssilvert** [comment]
I'm taking this out of triage, but we will address it in login V2.

**Comment by edewit** [comment]
We are going to replace all the configurable class names in v2 with "hardcoded" once and rely on patternfly variables to customize as we already do for admin and account

**Comment by garronej** [comment]
Hello @edewit,

Customizing the PatternFly CSS variables is a great option, but the current system also allows for incrementally overwriting the PatternFly styles using other styling solutions, which provides even more flexibility. In practice, the majority of users prefer Tailwind.

Here is the approach that Keycloakify users implement:

	•	Incrementally disabling the default styles: [Keycloakify Documentation](https://docs.keycloakify.dev/customization-strategies/css-level-customization/removing-the-default-styles)
	•	Tailwind example: [Keycloakify Documentation](https://docs.keycloakify.dev/customization-strategies/css-level-customization/using-tailwind)

What I appreciate about the current system is that Keycloakify can fully mirror the default theme, as the components are just a React translation of the FreeMarker templates. If you decide to remove the standardized class targets, Keycloakify would need to introduce its own. While this isn’t a major issue, I like the idea of Keycloakify providing integration tooling around what you are building rather than offering an alternative solution. This change would lead Keycloakify to diverge slightly from the default theme.

**Comment by dasniko** [comment]
I'm neither of the opinion that today most of the companies want to use something like Keycloakify (no offense, I really like the project!), nor that they are using mostly Tailwind.

I did a lot of frontend stuff in the recent years and still helping my customers to implement Keycloak themes. None of them is using Patternfly at all, most of them doesn't even know Patternfly _at all_.

I see more something like this: One third is using Keycloakify, one third is using something Bootstrap-based and the last third is using a complete custom CSS library/sheet/solution/whatever.
For the second and third thirds, the current (v1) approach is quite nice, as most of the time only a bit of HTML adjustment is necessary (mostly the `template.ftl`) and nearly all of the CSS classes can be set through `template.properties` file. Only the few hardcoded classnames are annoying and there are some nodes/divs missing a `class` attribute at all.

Going with hardcoded PF-classnames in the `v2` login theme makes it very hard to customize the various pages to the companies CI/CD. Editing CSS variables is not enough, especially if the custom approach doesn't use PF at all.
That's a bummer, if you really decide to go this way. Just my two cents.

**Comment by garronej** [comment]
@dasniko, I fully agree with your points about Patternfly and the class attribute limitations. Your perspective on Keycloakify suggests I should clarify its purpose and functionality.

It seems you believe Keycloakify is primarily for taking ownership of page structures and code with React. In reality, we now recommend maximizing customization through styling and only resorting to code modifications when absolutely necessary.

Keycloakify effectively covers all three approaches you mentioned. Its core features include:

- Previewing your pages in Storybook. [docs](https://docs.keycloakify.dev/testing-your-theme/in-storybook)
- Testing in a local Keycloak container with hot reloading enabled via a simple command. [docs](https://docs.keycloakify.dev/testing-your-theme/in-a-keycloak-docker-container)  
- Bundling your theme into a .jar file.

Whether you use Bootstrap or any other library/style solution, Keycloakify simplifies the process.Beyond that, any existing existing Keycloak theme into a Keycloakify theme in under five minutes, provided the Freemarker templates haven't been modified, see [this guide](https://docs.keycloakify.dev/customization-strategies/css-level-customization/applying-your-own-classes).

Keycloakify also ensures your theme is compatible with Keycloak versions 11 to 25 by polyfilling the newer Keycloak APIs and features client-side validation of user profile input fields.

Customizing components with React is an added bonus if you want to use libraries like MUI or Shadcn/UI.

In short, while I may be biased, I believe Keycloakify is the best approach in all scenarios except when you prefer not to introduce a third-party tool into your stack (which I can understand).

Here's a recent video I recorded about creating theme variants: [Watch here](https://youtu.be/Nkoz1iD-HOA?si=PniolrkzhrBHg0AQ).

**Comment by dasniko** [comment]
@garronej This sounds great! Yes, I might have not enough knowhow about recent Keycloakify features, as I have enough to dos with other Keycloak stuff... 😉  Perhaps it's worth giving it another try. Or, we should do a video togehter on my channel...!? I'll send you an email...!

**Comment by garronej** [comment]
> Or, we should do a video togehter on my channel...!?  

I'd love that! Let's get in touch!

On a side note, you might be right about Tailwind. I could be overestimating the number of people using it because of the many questions I receive about it on Discord.

**Comment by garronej** [comment]
And lastly, when I say using Talwind it doen's mean that pepole have to directly includes the tw utilities classes in the component structure, they can if they want but the recommended approach is to use classes attribute targets:    
Example:  

https://github.com/keycloakify/keycloakify-starter/blob/dd516e53e4dfa7c1ce02bab557420b999e87eca2/src/login/index.css

**Comment by garronej** [comment]
@edewit @jonkoops If you'll allow me, here are three suggestions that could greatly benefit the Keycloakify community of users moving forward:

- **Option 1:** Move on with your project of hardcoding the classes while maintaining standardized, style-free classes. This would allow users to apply their own styles, for example:

```ftl
<ul class="kcFormSocialAccountListClass pf-c-login__main-footer-links kc-social-links">...</ul>
```

This approach would require significant effort on the Keycloakify side but we could make it work.

- **Option 2:** Keep the current system and consider adding more class properties.

- **Option 3:** It might be bold of me to suggest this, but have you considered adopting Keycloakify?

Here are the arguments in favor of adopting it:

- **User Profile Attribute Validators:** Keycloakify implements these validators exposed to the FreeMarker context, enabling client-side validation of inputs for more immediate feedback.
- **Theme Development Continuity:** You can continue developing the user facing pages without worrying about breaking existing user themes.
- **Type Safety:** Full type safety for the FTL context, i18n, and class names, which should allow you to develop more quickly and with less concern about unintentionally breaking things.
- **Immediate Updates:** Instead of us playing catch-up to integrate new features, Keycloakify users would receive updates as they are released.
- **One Less Thing to Think About:** You wouldn't have to worry about creating tooling and documentation for customizing the theme anymore.
- **Development and Testing Tools:** You’d be able to develop your pages on Storybook [Example](https://storybook.keycloakify.dev/). Implement [visual regression testing](https://storybook.js.org/docs/writing-tests/visual-testing), and showcase the user-facing pages in all their aspects.  
- **Minimal Overhead:** Keycloakify adds less than a 10% increase to the page payload size compared to the default theme. Considering this extra JavaScript also enables client-side validation, the trade-off is worthwhile.
- **Monorepo Integration Support:** Easy to integrate into your workflow.
- [Keycloakify is self-contained](https://www.npmjs.com/package/keycloakify/v/10.0.0-rc.120?activeTab=dependencies); it has no third-party dependencies like i18next or react-hook-form. It won’t add thousands of dependencies to your node_modules directory.

If, however, none of these options are selected, it would imply too much work for us to keep mirroring the default theme one-to-one while at the same time maintaining the level of style customization Keycloakify users are accustomed to. Consequently, we would stop using PatternFly and rewrite the entire template using Tailwind CSS and [ShadeCN/UI](https://ui.shadcn.com/examples/authentication) instead.

Let me know if you’re interested in trying out Keycloakify—I can prepare a PR to set everything up, starting with the login theme.

**Comment by edewit** [comment]
I'm sorry, but I don't see the problem that these "hardcoded" class names have. What difference does it make if the class is called `pf-v5-c-form-control` or you set it in a property you need to create a custom css in both cases? Also @dasniko  I'm really interested in knowing what can't be achieved by changing the PF variables? What we don't want is having 3 different ways to style things and for login to work differently then the others.

**Comment by jonkoops** [comment]
Just be aware that any future PatternFly upgrades will have new class names. So `pf-v5-c-form-control` will have to be converted to `pf-v6-c-form-control` when we upgrade.

**Comment by garronej** [comment]
> What difference does it make if the class is called `pf-v5-c-form-control` or you set it in a property you need to create a custom css in both cases?  

The primary distinction is that these classes have predefined styles applied to them. There is no straightforward method to selectively disable these default styles.  
The only option is to cease importing the PatternFly CSS, which necessitates a complete re-styling effort. Moreover, targeting pf-something classes while employing a different design system seems inappropriate.  
Additionally, this method is incompatible with CSS frameworks that utilize utility classes, such as Bootstrap, Bulma, or Foundation.  
Finally, as @jonkoops pointed out, these classes lack stability.

**Comment by jonkoops** [comment]
My vote would be for option number 2, which is to get rid of hard-coded classes as much as possible and use the properties file, as the original theme did as well. This allows for easy customization, especially for users who are interested in just overriding some things here and there.

I think also that we should 'componentify' the templates more so that pieces can be replaced in a more modular format. For example, a button could live in it's own template and have it's logic constrained there so that it can be re-used in various places in other templates. This makes it trivial for a user that wants to customize the look and feel of only select parts of the UI.

As for adopting Keycloakify, I think we should attempt to take from it what makes sense and upstream it, this would also reduce the maintenance burden for the community. @garronej if you know of any features that would make sense to upstream here I'd be interested in what you have in mind.

**Comment by garronej** [comment]
> My vote would be for option number 2

Great! That works for us!

> I think also that we should 'componentify' the templates more

This would be a **big** step forward! Allowing people to "eject" components like buttons and replace them everywhere would make a huge difference. Currently, they can only eject the common template, the pages, and the user profiles form. However, most people just want to use their own base MUI/TailwindUI/Shadecn components without taking ownership of excessive code.

> If you know of any features that would make sense to upstream here I'd be interested in what you have in mind.

Well, we have over three years of active development, so we have a lot of features:  
- Full type safety on the Keycloak context/i18n keys and standardized class names.
- Users can see what's in the Freemarker context in the developer tool.
- The ability to test the pages without having to reproduce the flow in a real Keycloak.
- The ability to test the theme on any version of Keycloak with a simple command.
- Access to environment variables to enable customizability without having to rebuild.
- The ability to create theme variants.
- Real-time validation of the User Profile Attributes and password policies.
- A backward compatibility layer is provided, so all Keycloak versions are supported.
- The ability to use modern bundlers, allowing people to use whatever web technology they want in a familiar way.
- The ability to bundle additional Keycloak extensions in their theme.
- If they want to customize a page at the component level, there's an interactive CLI command to assist in this process, and the components use React instead of FreeMarker.

However, most of these features are only made possible by the Keycloakify bundler and aren't easily upstreamable with the current paradigm. It's alright; we can live side by side, but we could go further if we worked on the same baseline. I'm convinced that adopting Keycloakify would skyrocket the development of user-facing pages for you and be a game changer on how easy it is to build custom UX for Keycloak users.

**Comment by dasniko** [comment]
@garronej and @jonkoops already mentioned the most important parts. Thanks for that. 🙏

To put it a bit more boldly:

In PF, you have e.g. (currently) `pf-c-v5-form-control`, in another CSS library, e.g. Bootstrap, you have `form-control`. If you hard code every class in the `.ftl` files, one would have to overwrite _all_ `.ftl` files to customize the layout. Bad.

Adjusting PF variables means that one would have to have deep knowledge about the PF framework and also only adjusting existing classes is most often not enough and does not lead to the desired (or required) layout, as it still looks like PF. As already mentioned, at least among my customers (these are quite a few), nobody knows PF at all.

Additionally, the versioning issue, like @jonkoops mentioned.

> What we don't want is having 3 different ways to style things and for login to work differently then the others.

Understandable, but one of the "others" is the admin console. Adjusting this is (currently) a nightmare (cloning the project to be able to adjust it in a meaningful manner!) and most often, adjusting an admin UI is not needed at all. The other "others" is the account console, which is... well, at least "something" one _can_ use, but the fewest _want_ this to use and write their own self-service app, with proper CIAM possibilities.
So, there's left the login theme, which has a working approach currently. Changing this entirely is not understandable.
There's one more left - the email theme. This currently works also the same way with `.ftl` files like the login theme. The only difference is, that there's no default HTML/CSS layout for the emails, but companies do also style their emails!

**Comment by garronej** [comment]
@dasniko **YES**. Very well put.

**Comment by edewit** [comment]
This one works towards it https://github.com/keycloak/keycloak/pull/31793

**Comment by edewit** [comment]
closing done
